packages/backend-defaults/src/entrypoints/rootHttpRouter/http/RateLimitStoreFactory.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Config } from '@backstage/config';
+import type { Store } from 'express-rate-limit';
+import KeyvRedis from '@keyv/redis';
+import { RedisStore } from 'rate-limit-redis';
+
+/**
+ * Creates a store for `express-rate-limit` based on the configuration.
+ *
+ * @public
+ */
+export class RateLimitStoreFactory {
+  constructor(private readonly config: Config) {}
+
+  create(): Store | undefined {
+    const storeType = this.config.getOptionalString('backend.rateLimit.store');
+    if (!storeType) {
+      return this.auto();
+    }
+    switch (storeType) {
+      case 'redis':
+        return this.redis();
+      default:
+        throw new Error(
+          `Invalid 'backend.rateLimit.store' provided: ${storeType}`,
+        );
+    }
+  }
+
+  private auto(): Store | undefined {
+    const cacheStore =
+      this.config.getOptionalString('backend.cache.store') || 'memory';
+    // Use redis as primary if available
+    if (cacheStore === 'redis') {
+      return this.redis();
+    }
+
+    // Fallback to undefined (memory)
+    return undefined;
+  }
+
+  redis(): Store {
+    const connectionString =
+      this.config.getOptionalString('backend.cache.connection') || '';
```

Comment:
Does this make sense to be possible to leave empty? 🤔 I presume that keyv applies some default then? It's not optional [in `backend.cache`](https://github.com/backstage/backstage/blob/5bff622684aef50fc03c02c95eae878cd2006a03/packages/backend-defaults/config.d.ts#L486). Note also how that one uses a discriminated union for different connection types - could our config here also be grouped up like

```ts
backend?: {
  rateLimit?: false | {. // leveraging false to disable it all instead of separate flag
    connection:
      | {
          client: 'redis',
          // redis specific options here
        }
      | {
          // ...
        }
      | undefined,
    options: {
      // grouping common rate limiting settings here
    }
```

Mildly related, is it worth calling this `incomingRequestRateLimit` to be sure that it's clear what type of limiting that's intended? Especially since we're currently looking into outgoing request limiters to deal with exhaustion when interacting with e.g. github.